### ISSUE TITLE
Tailwind CSS v4.0 support

### ISSUE THREAD

**Issue by rickyzhangca** [adoxography/tailwind-scrollbar#104]
with the Tailwind CSS v4.0 beta, this error is given regardless of using the config file or the new CSS-first approach to import this plugin, appreciate anyone could take a look!

```
[plugin:@tailwindcss/vite:generate:serve] Package subpath './lib/util/flattenColorPalette' is not defined by "exports" in /Users/.../Documents/.../node_modules/.pnpm/tailwind-scrollbar@3.1.0_tailwindcss@4.0.0-beta.1/node_modules/tailwindcss/package.json
```

**Comment by rickyzhangca** [comment]
https://github.com/tailwindlabs/tailwindcss/pull/15318 this should fix the bug

**Comment by mst-mkt** [comment]
> [tailwindlabs/tailwindcss#15318](https://github.com/tailwindlabs/tailwindcss/pull/15318) this should fix the bug, needs to confirm once it's out

This error is resolved at `tailwindcss@4.0.0-beta.7`, but at the same time the following different errors occur

```err
Package subpath './lib/util/toColorValue' is not defined by "exports" in repo/node_modules/.pnpm/tailwind-scrollbar@3.1.0_tailwindcss@4.0.0-beta.7/node_modules/tailwindcss/package.json
```

This `toColorValue` function does not exist in the tailwind v4 code, so it may need to be addressed on the `tailwind-scrollbar` side

**Comment by mst-mkt** [comment]
`toColorValue` is the following function. If the argument is a string, it is considered to work without this function.
```ts
export default function toColorValue(maybeFunction) { 
   return typeof maybeFunction === 'function' ? maybeFunction({}) : maybeFunction 
} 
```
The following two locations are used in this library.
Since both pass strings, they probably do not need to be used.

1. https://github.com/adoxography/tailwind-scrollbar/blob/42cae713a0fe3e1aae8dd38907f9193b9441d443/src/utilities.js#L125-L128
In this case, there is no difference between `colors` and `themeColors` because [`flattenColorPalette` returns `Record<string, string>`](https://github.com/tailwindlabs/tailwindcss/blob/next/packages/tailwindcss/src/compat/flatten-color-palette.ts#L8-L29).

2. https://github.com/adoxography/tailwind-scrollbar/blob/42cae713a0fe3e1aae8dd38907f9193b9441d443/src/utilities.js#L130-L134
In this case, there is no difference between `color` and `value` because the object of the [first argument of `matchUtilities` has the value of type `(value: string, extra: { modifier: string | null }) => CssInJs | CssInJs[]`](https://github.com/tailwindlabs/tailwindcss/blob/next/packages/tailwindcss/src/compat/plugin-api.ts#L54-L58).

**Comment by rickyzhangca** [comment]
thanks!

**Comment by adoxography** [comment]
So, for full transparency here, I'm going to drag my heels on officially supporting v4 until I can get the package's test suite running using their new architecture. The way Tailwind is actually invoked has changed in v4 (i.e. it's not just a `postcss` plugin anymore), and that means that none of the tests are presently able to run.

Although it may seem like a simple little plugin, it's shockingly difficult to target pseudoclasses of pseudoelements (think `::-webkit-scrollbar-thumb:hover`) in Tailwind. As a result, this plugin has to do some manipulation on the `postcss` AST that Tailwind generates, and that's a little dangerous; it's easy to create scenarios like #101 where that manipulation inadvertently messes up the CSS output from completely unrelated packages. I have tests in the test suite that should prevent any regressions, and I want confirmation from them before I go about releasing anything.

I'll finally have some down time at work this month, so I'm hoping to find time to do some reading on how the authors intend Tailwind to be invoked now

**Comment by adoxography** [comment]
I've released a version that's compatible with Tailwind 4 as `tailwind-scrollbar@4.0.0-beta.0`. The following **breaking changes** apply only to it, not the 3.x.x releases.

As I feared, the (admittedly shady) system I was using to get `hover:` and `active:` to work was rendered ineffective by the new compiler Tailwind is using. This means that a class like `hover:scrollbar-thumb-red-100` will revert to being applied when the **element** is hovered, not the scrollbar's thumb. Most of the time, that's not what you want - although I do like that this logic is now at least possible (if unusual).

In its place, I've added a couple of dedicated variants, `scrollbar-hover:` and `scrollbar-active:`. These really mean "scrollbar thumb hovered/active," but it's usually the thumb you want to target with these variants. If for some reason you really do want to target the track or corner's hover/active states, I've also added `scrollbar-track-hover:`, `scrollbar-track-active:`, `scrollbar-corner-hover:`, and `scrollbar-corner-active:` (but if you use them, I will think you are a weirdo).

**tl;dr**: Use `tailwind-scrollbar@4.0.0-beta.0` for compatibility with Tailwind 4. Do a global find and replace for `hover:scrollbar` to `scrollbar-hover:scrollbar` and `active:scrollbar` to `scrollbar-active:scrollbar`.

**Comment by aimz-yeah** [comment]
NIce, was this beta ready for release? Would it just be a case of updating the peer dependancies from 3.x to allow this to work with Tailwind 4 now that it's been released out of beta?

**Comment by bicarlsen** [comment]
I **don't** use `hover:scrollbar` or `active:scrollbar`. After bumping to `tailwind@4.0.0` and `tailwind-scrollbar@4.0.0-beta.0` everything is working.

Thanks!

**Comment by adoxography** [comment]
Yup, beta was ready for release - just needed to update the docs to reflect the points I made above. I've tagged `4.0.0` as latest, so you should be good to go with Tailwind 4.
